rig/place_and_route/place/rcm.py

```diff
@@ -0,0 +1,178 @@
+        _visited = set()
+
+    yield vertex
+    _visited.add(vertex)
+
+    for neighbour in vertices_neighbours[vertex]:
+        if neighbour not in _visited:
+            for vertex in _dfs(neighbour, vertices_neighbours, _visited):
+                yield vertex
+
+
+def _get_connected_subgraphs(vertices, vertices_neighbours):
+    """Break a graph containing unconnected subgraphs into a list of connected
+    subgraphs.
+
+    Returns
+    -------
+    [set([vertex, ...]), ...]
+    """
+    remaining_vertices = set(vertices)
+    subgraphs = []
+    while remaining_vertices:
+        subgraph = set(_dfs(remaining_vertices.pop(), vertices_neighbours))
+        remaining_vertices.difference_update(subgraph)
+        subgraphs.append(subgraph)
+
+    return subgraphs
+
+
+def _cuthill_mckee(vertices, vertices_neighbours):
+    """Yield the Cuthill-McKee order for a connected, undirected graph.
+
+    `Wikipedia
+    <https://en.wikipedia.org/wiki/Cuthill%E2%80%93McKee_algorithm>`_ provides
+    a good introduction to the Cuthill-McKee algorithm. The RCM algorithm
+    attempts to order vertices in a graph such that their adjacency matrix's
+    bandwidth is reduced. In brief the RCM algorithm is a breadth-first search
+    with the following tweaks:
+
+    * The search starts from the vertex with the lowest degree.
+    * Vertices discovered in each layer of the search are sorted by ascending
+      order of their degree in the output.
+
+    .. warning::
+
+        This function must not be called on a disconnected or empty graph.
+
+    Returns
+    -------
+    [vertex, ...]
+    """
+    vertices_degrees = {v: sum(itervalues(vertices_neighbours[v]))
+                        for v in vertices}
+
+    peripheral_vertex = min(vertices, key=(lambda v: vertices_degrees[v]))
+
+    visited = set([peripheral_vertex])
+    cm_order = [peripheral_vertex]
+    previous_layer = set([peripheral_vertex])
+    while len(cm_order) < len(vertices):
+        adjacent = set()
+        for vertex in previous_layer:
+            adjacent.update(vertices_neighbours[vertex])
+        adjacent.difference_update(visited)
+
+        visited.update(adjacent)
+        cm_order.extend(sorted(adjacent, key=(lambda v: vertices_degrees[v])))
+        previous_layer = adjacent
+
+    return cm_order
+
+
+def rcm_vertex_order(vertices_resources, nets):
+    """A generator which iterates over the vertices in Reverse-Cuthill-McKee
+    order.
+
+    For use as a vertex ordering for the sequential placer.
+    """
+    vertices_neighbours = _get_vertices_neighbours(nets)
+    for subgraph_vertices in _get_connected_subgraphs(vertices_resources,
+                                                      vertices_neighbours):
+        cm_order = _cuthill_mckee(subgraph_vertices, vertices_neighbours)
+        for vertex in reversed(cm_order):
+            yield vertex
+
+
+def rcm_chip_order(machine):
+    """A generator which iterates over a set of chips in a machine in
```

Comment:
You may be being misled: the implementation internally calls `rcm_vertex_order` to generate the ordering however it generates a graph in which the vertices are `(x, y)` chip coordinates and the nets represent working connections between those chips. As a result, the ordering produced by `rcm_vertex_order` is actually an ordering of chip coordinates.

Does that make sense? (What would be a better way of making this clearer in the code?)